### DISCUSSION TITLE
Problem for creating plugin

### DISCUSSION THREAD

**Discussion by ssogari-dev** [streamlink/streamlink#5733] [Q&A] [answered]
### Description

hello,
I'm a beginner developer making a plugin for Streamlink in Korea.

I apologize for the unavoidable use of machine translation.

Recently, a new streaming service called 'NAVER CHZZK' was launched in Korea, and I am creating a streamlink plugin that can be used on this service. The developer of Streamlink declined my request because the service is still in beta, but since the number of users is increasing, I decided to create this plugin by myself.

The user enters a URL of the form below.
https://chzzk.naver.com/live/{channel_id}

The plugin then gets the information from the API below.
https://api.chzzk.naver.com/service/v1/channels/{channel_id}/live-detail

I wrote code to get both the broadcast information and the HLS link from the above API. It looks like this. (Retrieving broadcast information and HLS URLs via the API is no problem.)
```python
import re, json, requests
# import m3u8
from streamlink.plugin import Plugin, pluginmatcher
from streamlink.stream import HLSStream

@pluginmatcher(re.compile(r'https://chzzk\.naver\.com/live/(?P<channel_id>\w+)'))
class ChzzkPlugin(Plugin):
    API_URL = "https://api.chzzk.naver.com/service/v1/channels/{channel_id}/live-detail"

    def _get_streams(self):
        channel_id = self.match.group("channel_id")
        api_url = self.API_URL.format(channel_id=channel_id)

        try:
            response = requests.get(api_url)
            response.raise_for_status()
        except requests.RequestException as e:
            self.logger.error("Failed to fetch channel information: {0}".format(str(e)))
            return

        if response.status_code == 404:
            self.logger.error("Channel not found")
            return

        try:
            content = response.json().get('content', {})
            status = content.get('status')
            if status != 'OPEN':
                self.logger.error("Channel is not live (status: {0})".format(status))
                return

            live_title = content.get('liveTitle')
            channel_name = content.get('channel', {}).get('channelName')
            category = content.get('liveCategory')
            stream_info = content.get('livePlaybackJson')
            hls_url = json.loads(stream_info).get('media', [{}])[0].get('path')
            # hls = m3u8.loads(hls_url)
            
            self.logger.info("Stream Title: {0}".format(live_title))
            self.logger.info("Channel Name: {0}".format(channel_name))
            self.logger.info("Category: {0}".format(category))
            self.logger.info("HLS URL: {0}".format(hls_url))

            # Use HLS URL directly to create HLSStream
            streams = {
                "live": HLSStream(self.session, hls_url),
            }

            return streams
        except json.JSONDecodeError as e:
            self.logger.error("Failed to decode JSON response: {0}".format(str(e)))
            return

__plugin__ = ChzzkPlugin
```

But here's where we run into a problem.
I get the HLS link fine, but when I pass it to Streamlink's HLSstream, it doesn't work correctly.

However, when I put the resolved HLS URL back into streamlink based on the link entered by the user, it works as an HLS plugin and displays correctly.

I was wondering how I can pass the verified HLS link back to the HLSstream when I put the original CHZZK link when using the output function.

Also, if it is possible, I was wondering if there is a way to get the metadata of the broadcast information (broadcast title, channel name, category, etc). Like in the case of Twitch, when you type {title}, the broadcast title comes up.

**+ as reminder**
1. there is a separate HLS link for low latency. This will be added later when I modify the code (so that it works when I receive the --chzzk-low-latency option as an argument).
2. one m3u8 link contains multiple quality videos all in one. It probably looks similar to Twitch, so there shouldn't be a problem.

**Comment by skulblakka** [comment, marked as answer]
To get it to work you can simply replace the bottom part of your `_get_streams` function with the following:
```python
            self.logger.info("Stream Title: {0}".format(live_title))
            self.logger.info("Channel Name: {0}".format(channel_name))
            self.logger.info("Category: {0}".format(category))
            self.logger.info("HLS URL: {0}".format(hls_url))
            self.author = channel_name
            self.category = category
            self.title = live_title

            # Use HLS URL directly to create HLSStream
            yield from HLSStream.parse_variant_playlist(self.session, hls_url).items()
        except json.JSONDecodeError as e:
            self.logger.error("Failed to decode JSON response: {0}".format(str(e)))
            return
```

This will give you streamer, category, title and the stream in the qualities available.

```bash
streamlink https://chzzk.naver.com/live/7ce8032370ac5121dcabce7bad375ced best --title "{author} - {category} - {title}"
[cli][info] Found matching plugin chzzk for URL https://chzzk.naver.com/live/7ce8032370ac5121dcabce7bad375ced
[plugins.chzzk][info] Stream Title: 반반유치원6 공포
[plugins.chzzk][info] Channel Name: 풍월량월풍월량
[plugins.chzzk][info] Category: talk
[plugins.chzzk][info] HLS URL: https://livecloud.akamaized.net/chzzk/lip2_kr/pgsg1nmss2u0002/xk9zjn5k0tqup5sbecaltncaf8isk5bdcs/hls_playlist.m3u8?hdnts=st=1703094381~exp=1703126791~acl=*/xk9zjn5k0tqup5sbecaltncaf8isk5bdcs/*~hmac=4a548cf03f3393cc396125de78bd156b9ceb583e2058a6b85b4549cbcac7b963
[cli][info] Available streams: 144p (worst), 360p, 480p, 720p, 1080p (best)
[cli][info] Opening stream: 1080p (hls-multi)
[cli][info] Starting player: H:\Tresors\Dev\mpv\build\mpv.exe
[utils.named_pipe][info] Creating pipe streamlinkpipe-28272-1-9276
[utils.named_pipe][info] Creating pipe streamlinkpipe-28272-2-2180
[cli][info] Player closed
[cli][info] Stream ended
[cli][info] Closing currently open stream...
```

**Reply by ssogari-dev** [reply]
Awesome! Thank you so much for the answer.
You've made my plugin creation a lot easier.

I've posted the first finished version on [my repo](https://github.com/ssogari-dev/Chzzk_Tool/blob/main/NaverChzzk.py). 

Have a great day!
